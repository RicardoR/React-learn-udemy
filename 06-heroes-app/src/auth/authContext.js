import { createContext } from 'react';

// Uppercase because its a component
export const AuthContext = createContext();
