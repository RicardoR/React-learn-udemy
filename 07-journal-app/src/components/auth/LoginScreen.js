import React, { useState } from 'react';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import {
  startLoginWithEmailPassword,
  startGoogleLogin,
} from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {
  const dispatch = useDispatch();
  const [error, setError] = useState(undefined);
  const { msgError, loading } = useSelector((state) => state.ui);

  const [formValues, handleInputChange] = useForm({
    email: 'email@gmail.com',
    password: '123456',
  });

  const { email, password } = formValues;

  const handleLogin = (event) => {
    event.preventDefault();
    if (isFormValid()) {
      dispatch(startLoginWithEmailPassword(email, password));
    }
  };

  const handleGoogleLogin = () => {
    if (!loading) {
      dispatch(startGoogleLogin());
    }
  };

  const isFormValid = () => {
    if (!email || !password) {
      setError('All fields are required');
      return false;
    }

    if (!validator.isEmail(email)) {
      setError('Invalid email');
      return false;
    }

    if (password.length < 5) {
      setError('Password must be at least 5 characters');
      return false;
    }

    setError(undefined);
    return true;
  };

  return (
    <>
      <h3 className="auth__title mb-5">Login</h3>
      <form onSubmit={handleLogin}>
        {(error || msgError) && (
          <p className="alert-error mb-5">{error || msgError}</p>
        )}
        <input
          type="text"
          placeholder="Email"
          name="email"
          className="auth__input"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="password"
          name="password"
          className="auth__input"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <button
          disabled={loading}
          type="submit"
          className="btn btn-primary btn-block mb-5 mt-1"
        >
          Login
        </button>
        <div>
          <p className="mb-5">Login with social networks</p>

          <div className="google-btn mb-5" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>
        <Link className="link" to="/auth/register">
          Create new account
        </Link>
      </form>
    </>
  );
};
