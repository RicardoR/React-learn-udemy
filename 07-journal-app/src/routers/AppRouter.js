import React from 'react';
import { Redirect } from 'react-router-dom';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import { JournalScreen } from '../components/journal/JournalScreen';
import { AuthRouter } from './AuthRouter';

export const AppRouter = () => {
  return (
    <Router>
      <div>
        <Switch>
          <Route path="/auth">
            <AuthRouter />
          </Route>
          <Route path="/" exact>
            <JournalScreen />
          </Route>
          <Redirect to="/auth/login" />
        </Switch>
      </div>
    </Router>
  );
};
