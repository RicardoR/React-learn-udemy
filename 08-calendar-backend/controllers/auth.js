const { response } = require('express');

const createUser = (request, resp = response) => {
  const { name, email, password } = request.body;

  if (name.length < 5) {
    return resp.status(400).json({
      ok: false,
      msg: 'El nombre debe tener 5 caracteres',
    });
  }

  resp.status(201).json({
    ok: true,
    msg: 'registro',
    name,
    email,
    password,
  });
};

const performLogin = (request, resp = response) => {
  const { email, password } = request.body;

  resp.json({
    ok: true,
    msg: 'login',
    email,
    password,
  });
};

const renewToken = (request, resp = response) => {
  const { token } = request.body;

  resp.json({
    ok: true,
    msg: 'renew',
    token,
  });
};

module.exports = { createUser, performLogin, renewToken };
