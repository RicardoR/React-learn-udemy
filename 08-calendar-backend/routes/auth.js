/*
    Auth routes  / Auth
    host + /api/auth
*/

const { Router } = require('express');
const { check } = require('express-validator');
const { fieldValidator } = require('../middlewares/field-validator');

const { createUser, performLogin, renewToken } = require('../controllers/auth');

const router = Router();

router.post(
  '/new',
  [
    // middelwares
    check('name', 'Name is required').not().isEmpty(),
    check('email', 'Email is required').isEmail(),
    check('password', 'Password should contain 6 characters').isLength({
      min: 6,
    }),
    fieldValidator,
  ],
  createUser
);

router.post(
  '/',
  [
    check('email', 'Email is required').isEmail(),
    check('password', 'Password should contain 6 characters').isLength({
      min: 6,
    }),
    fieldValidator,
  ],
  performLogin
);

router.get('/renew', renewToken);

module.exports = router;
