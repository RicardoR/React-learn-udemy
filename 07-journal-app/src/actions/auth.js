import { getAuth, signInWithPopup } from 'firebase/auth';

import { actionTypes } from '../types/actionTypes';
import { googleAuthProvider } from '../firebase/firebase-config';

export const initLogin = (email, password) => {
  return (dispatch) => {
    setTimeout(() => {
      dispatch(login(123, 'pedro'));
    }, 3500);
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    const auth = getAuth();
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName));
    });
  };
};

export const login = (uid, displayName) => {
  return {
    type: actionTypes.LOGIN,
    payload: {
      uid,
      displayName,
    },
  };
};
