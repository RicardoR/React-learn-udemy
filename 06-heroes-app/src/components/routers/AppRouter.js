import React from 'react';
import { Routes, Route, BrowserRouter } from 'react-router-dom';
import { DcScreen } from '../dc/DcScreen';
import { LoginScreen } from '../login/LoginScreen';
import { MarvelScreen } from '../marvel/MarvelScreen';
import { SearchScreen } from '../search/SearchScreen';
import { Navbar } from '../ui/Navbar';

export const AppRouter = () => {
  return (
    <BrowserRouter>
      <Navbar />
      <Routes>
        <Route path="/" element={<MarvelScreen />} />
        <Route path="/marvel" element={<MarvelScreen />} />
        <Route path="/dc" element={<DcScreen />} />
        <Route path="/login" element={<LoginScreen />} />
        <Route path="/search" element={<SearchScreen />} />
      </Routes>
    </BrowserRouter>
  );
};
