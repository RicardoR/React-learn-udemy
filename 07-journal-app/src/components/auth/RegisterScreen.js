import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startRegisterWithEmailAndPassword } from '../../actions/auth';

import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {
  // const [error, setError] = useState(undefined);

  const dispatch = useDispatch();
  const { msgError: error } = useSelector((state) => state.ui);

  const [formValues, handleInputChanges] = useForm({
    name: 'Ricardo',
    email: 'email@gmail.com',
    password: '123456',
    confirmPassword: '123456',
  });

  const { name, email, password, confirmPassword } = formValues;

  const handeRegister = (e) => {
    e.preventDefault();
    if (isFormValid()) {
      dispatch(startRegisterWithEmailAndPassword(email, password, name));
    }
  };

  const isFormValid = () => {
    if (name.trim().length === 0) {
      // setError('Name is required');
      dispatch(setError('Name is required'));
      return false;
    }

    if (validator.isEmail(email) === false) {
      // setError('Invalid email');
      dispatch(setError('Invalid email'));
      return false;
    }

    if (password.length < 5) {
      // setError('Password must be at least 5 characters');
      dispatch(setError('Password must be at least 5 characters'));
      return false;
    }

    if (password !== confirmPassword) {
      // setError('Passwords do not match');
      dispatch(setError('Passwords do not match'));
      return false;
    }

    // setError(undefined);
    dispatch(removeError());
    return true;
  };

  return (
    <>
      <h3 className="auth__title mb-5">Register</h3>
      <form onSubmit={handeRegister}>
        {error && <p className="alert-error mb-5">{error}</p>}
        <input
          type="text"
          placeholder="Name"
          name="name"
          className="auth__input"
          autoComplete="off"
          value={name}
          onChange={handleInputChanges}
        />
        <input
          type="email"
          placeholder="Email"
          name="email"
          className="auth__input"
          autoComplete="off"
          value={email}
          onChange={handleInputChanges}
        />
        <input
          type="password"
          placeholder="password"
          name="password"
          className="auth__input"
          autoComplete="off"
          value={password}
          onChange={handleInputChanges}
        />
        <input
          type="password"
          placeholder="Confirm password"
          name="confirmPassword"
          className="auth__input"
          autoComplete="off"
          value={confirmPassword}
          onChange={handleInputChanges}
        />
        <button type="submit" className="btn btn-primary btn-block mb-5 mt-1">
          Register
        </button>
        <div></div>
        <Link className="link" to="/auth/login">
          Already registered?
        </Link>
      </form>
    </>
  );
};
