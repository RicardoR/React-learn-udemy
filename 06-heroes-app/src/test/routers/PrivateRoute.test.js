import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { PrivateRoute } from '../../components/routers/PrivateRoute';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Navigate: () => <span>El login</span>,
}));

describe('PrivateRoute', () => {
  Storage.prototype.setItem = jest.fn();

  test('should display the component when logged', () => {
    const contextValue = {
      user: {
        logged: true,
        name: 'John Doe',
      },
    };
    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/']}>
          <PrivateRoute>
            <h1>Hello Private Component</h1>
          </PrivateRoute>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('h1').text().trim()).toBe('Hello Private Component');
    expect(Storage.prototype.setItem).toHaveBeenCalledTimes(1);
  });

  test('should navigate to login when no logged', () => {
    const contextValue = {
      user: {
        logged: false,
      },
    };
    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/']}>
          <PrivateRoute>
            <h1>Hello Private Component</h1>
          </PrivateRoute>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('span').text().trim()).toBe('El login');
  });
});
