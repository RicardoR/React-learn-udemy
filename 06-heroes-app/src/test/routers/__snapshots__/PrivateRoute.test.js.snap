// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute should display the component when logged 1`] = `ReactWrapper {}`;

exports[`PrivateRoute should navigate to login when no logged 1`] = `ReactWrapper {}`;
